### ISSUE TITLE
RFC: build images from every "work in progress" GitHub branch

### ISSUE THREAD

**Issue by ktdreyer** [ceph/ceph-container#651]
TL;DR: I would like us to build images from every "work in progress" GitHub branch and store them in a central registry.

### Background
There are currently two main build processes:

1. Sebastien runs a script to write the build-* GitHub branches and trigger image builds in https://hub.docker.com/r/ceph .
  The images built here can be consumed by other tools (manual testing, or teuthology, etc) or communities (eg. RDO, OpenShift). Users just need to `docker pull ceph/daemon` to start from the latest pre-built image.

2. Jenkins tests all Pull Requests.
   Jenkins runs `docker build` on all PRs and pushes the resulting image to a local ephemeral registry. For testing, Jenkins also deploys from that registry, then tears everything down at the end of the test.

### Drawbacks of the current Jenkins + Tox system:

The images built in this system can only be used from within the system. The "image-building" step is tightly coupled with the "testing" step. It is not easy to add more testing because everything is thrown away at the end (there is no central registry, and Mita/Jenkins wipes all the Jenkins slaves).

### Immediate goals:

- Automatically build container images for all "work in progress" branches of ceph-docker, not just master.

- Upload the results to a central registry for other test tools to consume.

These goals should allow us to add broader testing to ceph-docker artifacts. De-coupling the "image-building" steps from the "testing" steps makes it easier to implement long-running tests for containerized Ceph. Decoupling should also make it easier to share test code across ceph-docker upstream an RH's downstream products.

### Desired outcomes:

- Easier manual testing for things Tox does not cover.

- Open the possibility for Teuthology integration, so we get longer-running / more exhaustive tests of containerized Ceph clusters.

- Testing and integration with projects like RDO that want to consume Ceph in a container and want to get the very latest or experimental bits.

- Expose CI build artifacts to outside contributors that do not have access to jenkins.ceph.com infrastructure, but could still report pass/fail results for their own testing (like they do with packaged builds from shaman.ceph.com today).

### Progress so far:
Jenkins job to build containers, tag them as "ci", and then push them to the Docker Hub: https://github.com/ceph/ceph-build/pull/722 . This job is activated for testing, at https://jenkins.ceph.com/view/ceph-docker/ . It fails because jenkins.ceph.com does not have access to upload to the Docker Hub.

### Next steps:
The next step is to (re)create a "cephjenkins" Docker Hub account so jenkins.ceph.com can push the results to https://hub.docker.com/r/ceph.

When I discussed this with Sebastien today, I suggested we keep these images in an entirely separate repository, ie "ceph/daemon-ci" to indicate that they are fully automated and may not be as stable as the other images in "ceph/daemon".

### Longer-term goals (out of scope here):
These things would be great, but I treat them as out-of-scope for the immediate term:
- Dynamically manipulate the Dockerfile for these "CI image builds" to pull the latest Jewel RPMs from shaman.ceph.com, rather than pulling older RPMs from download.ceph.com.
- Expanding on the above bullet, dynamically inject repos for any ceph branch in shaman.ceph.com, not just the main "jewel", "kraken", etc. release branches.
